torchrl/data/tensor_specs.py

```diff
@@ -1014,61 +1027,77 @@ class MultiDiscreteTensorSpec(DiscreteTensorSpec):
 
     def __init__(
         self,
-        nvec: Sequence[int],
+        nvec: Union[Sequence[int], torch.Tensor],
         device: Optional[DEVICE_TYPING] = None,
         dtype: Optional[Union[str, torch.dtype]] = torch.long,
     ):
+        if not isinstance(nvec, torch.Tensor):
+            nvec = torch.tensor(nvec)
+        self.nvec = nvec
         dtype, device = _default_dtype_and_device(dtype, device)
-        self._size = len(nvec)
-        shape = torch.Size([self._size])
-        space = BoxList([DiscreteBox(n) for n in nvec])
+        shape = nvec.size()
+        space = BoxList.from_nvec(nvec)
         super(DiscreteTensorSpec, self).__init__(
             shape, space, device, dtype, domain="discrete"
         )
 
+    def _rand(self, space: Box, shape: torch.Size):
+        x = []
+        for _s in space:
+            if isinstance(_s, BoxList):
+                x.append(self._rand(_s, shape))
+            else:
+                x.append(
+                    torch.randint(
+                        0,
+                        _s.n,
+                        torch.Size([1, *shape]),
```

Comment:
I'm a bit puzzled, maybe there's something I'm missing
Can't we just do

```python
                x.append(
                    torch.randint(
                        0,
                        _s.n,
                        shape,
                        device=self.device,
                        dtype=self.dtype,
                    )
                )
        return torch.stack(x, -1)

```
and avoid the permutation in the end?